File: app_pojavlauncher/src/main/java/net/kdt/pojavlaunch/LauncherGLSurface.java
package net.kdt.pojavlaunch;

import static net.kdt.pojavlaunch.CallbackBridge.windowRate;
import static net.kdt.pojavlaunch.MainActivity.touchCharInput;
import static net.kdt.pojavlaunch.utils.MCOptionUtils.getMcScale;
import static net.kdt.pojavlaunch.CallbackBridge.sendMouseButton;
import static net.kdt.pojavlaunch.CallbackBridge.windowHeight;
import static net.kdt.pojavlaunch.CallbackBridge.windowWidth;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.os.Build;
import android.util.AttributeSet;
import android.util.Log;
import android.view.InputDevice;
import android.view.KeyEvent;
import android.view.MotionEvent;
import android.view.Surface;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.RequiresApi;

import net.kdt.pojavlaunch.customcontrols.ControlLayout;
import net.kdt.pojavlaunch.customcontrols.gamepad.DefaultDataProvider;
import net.kdt.pojavlaunch.customcontrols.gamepad.Gamepad;
import net.kdt.pojavlaunch.customcontrols.mouse.AndroidPointerCapture;
import net.kdt.pojavlaunch.customcontrols.mouse.InGUIEventProcessor;
import net.kdt.pojavlaunch.customcontrols.mouse.InGameEventProcessor;
import net.kdt.pojavlaunch.customcontrols.mouse.TouchEventProcessor;
import net.kdt.pojavlaunch.platform.input.PlatformGrabListener;
import net.kdt.pojavlaunch.platform.Platform;

import net.kdt.pojavlaunch.prefs.LauncherPreferences;
import net.kdt.pojavlaunch.render.SurfaceProvider;
import net.kdt.pojavlaunch.render.SurfaceViewSurfaceProvider;
import net.kdt.pojavlaunch.render.TextureViewSurfaceProvider;
import net.kdt.pojavlaunch.utils.MCOptionUtils;

import fr.spse.gamepad_remapper.GamepadHandler;
import fr.spse.gamepad_remapper.RemapperManager;
import fr.spse.gamepad_remapper.RemapperView;
import git.artdeell.dnbootstrap.glfw.GamepadEnableHandler;
import git.artdeell.mojoexec.MojoExec;

import static net.kdt.pojavlaunch.platform.Platform.PLATFORM;

/**
 * Class dealing with showing minecraft surface and taking inputs to dispatch them to minecraft
 */
public class LauncherGLSurface extends View implements PlatformGrabListener, GamepadEnableHandler, SurfaceProvider.SurfaceCallback {
    /* Gamepad object for gamepad inputs, instantiated on need */
    private GamepadHandler mGamepadHandler;
    /* The RemapperView.Builder object allows you to set which buttons to remap */
    private final RemapperManager mInputManager = new RemapperManager(getContext(), new RemapperView.Builder(null)
            .remapA(true)
            .remapB(true)
            .remapX(true)
            .remapY(true)

            .remapLeftJoystick(true)
            .remapRightJoystick(true)
            .remapStart(true)
            .remapSelect(true)
            .remapLeftShoulder(true)
            .remapRightShoulder(true)
            .remapLeftTrigger(true)
            .remapRightTrigger(true)
            .remapDpad(true));

    /* Sensitivity, adjusted according to screen size */
    private final double mSensitivityFactor = (1.4 * (1080f/ Tools.getDisplayMetrics((Activity) getContext()).heightPixels));

    private final SurfaceProvider mSurfaceProvider = LauncherPreferences.PREF_USE_ALTERNATE_SURFACE ? new SurfaceViewSurfaceProvider() : new TextureViewSurfaceProvider();
    private boolean mRefreshOnly = true;
    /* Surface ready listener, used by the activity to launch minecraft */
    SurfaceReadyListener mSurfaceReadyListener = null;
    final Object mSurfaceReadyListenerLock = new Object();
    /* View holding the surface, either a SurfaceView or a TextureView */
    View mSurface;

    private final InGameEventProcessor mIngameProcessor = new InGameEventProcessor(this, mSensitivityFactor);
    private final InGUIEventProcessor mInGUIProcessor = new InGUIEventProcessor(this);
    private TouchEventProcessor mCurrentTouchProcessor = mInGUIProcessor;
    private AndroidPointerCapture mPointerCapture;
    private View mTouchpad;
    private boolean mLastGrabState = false;

    public LauncherGLSurface(Context context) {
        this(context, null);
    }

    public LauncherGLSurface(Context context, AttributeSet attributeSet) {
        super(context, attributeSet);
        setFocusable(true);
        Platform.setGamepadEnableHandler(this);
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private void setUpPointerCapture() {
        if(mPointerCapture != null) mPointerCapture.detach();
        mPointerCapture = new AndroidPointerCapture(mTouchpad, this);
    }

    /** Initialize the view and all its settings
     * @param isAlreadyRunning set to true to tell the view that the game is already running
     *                         (only updates the window without calling the start listener)
     * @param touchpad the optional cursor-emulating touchpad, used for touch event processing
     *                 when the cursor is not grabbed
     */
    public void start(boolean isAlreadyRunning, View touchpad) {
        mTouchpad = touchpad;
        if (Tools.isAndroid8OrHigher()) setUpPointerCapture();
        mInGUIProcessor.setAbstractTouchpad(touchpad);
        mRefreshOnly = isAlreadyRunning;
        mSurface = mSurfaceProvider.create(getContext(), this);
        ((ViewGroup) getParent()).addView(mSurface);
    }

    /**
     * The touch event for both grabbed an non-grabbed mouse state on the touch screen
     * Does not cover the virtual mouse touchpad
     */
    @Override
    @SuppressWarnings("accessibility")
    public boolean onTouchEvent(MotionEvent e) {
        // Kinda need to send this back to the layout
        if(((ControlLayout)getParent()).getModifiable()) return false;

        // Looking for a mouse to handle, won't have an effect if no mouse exists.
        for (int i = 0; i < e.getPointerCount(); i++) {
            int toolType = e.getToolType(i);
            if(toolType == MotionEvent.TOOL_TYPE_MOUSE) {
                if(Tools.isAndroid8OrHigher() &&
                        mPointerCapture != null) {
                    mPointerCapture.handleAutomaticCapture();
                    return true;
                }
            }else if(toolType != MotionEvent.TOOL_TYPE_STYLUS) continue;

            // Mouse found
            // Avoid going through the JNI each time.
            if(Platform.isGrabbing()) return false;
            Platform.cursorX = e.getX(i) / getWidth();
            Platform.cursorY = e.getY(i) / getHeight();
            PLATFORM.sendMousePosition();
            return true; //mouse event handled successfully
        }
        if (mIngameProcessor == null || mInGUIProcessor == null) return true;
        boolean ret = mCurrentTouchProcessor.processTouchEvent(e);
        // Keep cursor on screen if panning with IME inset
        if(LauncherPreferences.PREF_KEYBOARD_AUTOPANNING && MainActivity.mImeHeight > 0){
            int translationY = Tools.getTranslationFromCursorY(
                    (int)(Platform.cursorY * mSurface.getHeight() + 100),
                    mSurface.getHeight(),
                    MainActivity.mImeHeight,
                    0
            );
            // If the view was force panned (KeyboardPan keycode) apply an animation instead of immediate override
            // This fixes weird jumps when the user moves the cursor first time after pressing that keycode
            if(MainActivity.mForceFullPanning) {
                mSurface.animate().setDuration(100).translationY(-translationY).start();
                mTouchpad.animate().setDuration(100).translationY(-translationY).start();
                MainActivity.mForceFullPanning = false;
            } else {
                mSurface.setTranslationY(-translationY);
                mTouchpad.setTranslationY(-translationY);
            }
        }
        return ret;
    }

    private void createGamepad(InputDevice inputDevice) {
        if(Platform.getPlatformGamepad() == null || Platform.getPlatformGamepad().shouldOverride())
            mGamepadHandler = new Gamepad(inputDevice, DefaultDataProvider.INSTANCE, mTouchpad);
    }

    /**
     * The event for mouse/joystick movements
     */
    @SuppressLint("NewApi")
    @Override
    public boolean dispatchGenericMotionEvent(MotionEvent event) {
        int mouseCursorIndex = -1;

        if(Gamepad.isGamepadEvent(event)){
            if(Platform.getPlatformGamepad() != null && Platform.getPlatformGamepad().shouldOverride()){
                Platform.getPlatformGamepad().sendMotionEvent(event);
                return true;
            }
            if(mGamepadHandler == null) createGamepad(event.getDevice());
            mInputManager.handleMotionEventInput(getContext(), event, mGamepadHandler);
            return true;
        }

        for(int i = 0; i < event.getPointerCount(); i++) {
            if(event.getToolType(i) != MotionEvent.TOOL_TYPE_MOUSE && event.getToolType(i) != MotionEvent.TOOL_TYPE_STYLUS ) continue;
            // Mouse found
            mouseCursorIndex = i;
            break;
        }
        if(mouseCursorIndex == -1) return false; // we cant consoom that, theres no mice!

        // Make sure we grabbed the mouse if necessary
        // Avoid going through the JNI each time.
        updateGrabState(Platform.isGrabbing());

        switch(event.getActionMasked()) {
            case MotionEvent.ACTION_HOVER_MOVE:
                Platform.cursorX = event.getX(mouseCursorIndex) / getWidth();
                Platform.cursorY = event.getY(mouseCursorIndex) / getHeight();
                PLATFORM.sendMousePosition();
                return true;
            case MotionEvent.ACTION_SCROLL:
                CallbackBridge.sendScroll(event.getAxisValue(MotionEvent.AXIS_HSCROLL), event.getAxisValue(MotionEvent.AXIS_VSCROLL));
                return true;
            case MotionEvent.ACTION_BUTTON_PRESS: sendMouseButton(event.getActionButton(),true); return true;
            case MotionEvent.ACTION_BUTTON_RELEASE: sendMouseButton(event.getActionButton(),false); return true;
            default:
                return false;
        }
    }

    /** The event for keyboard/ gamepad button inputs */
    public boolean processKeyEvent(KeyEvent event) {
        //Log.i("KeyEvent", event.toString());

        //Filtering useless events by order of probability
        int eventKeycode = event.getKeyCode();
        if(eventKeycode == KeyEvent.KEYCODE_UNKNOWN) return true;
        if(eventKeycode == KeyEvent.KEYCODE_VOLUME_DOWN) return false;
        if(eventKeycode == KeyEvent.KEYCODE_VOLUME_UP) return false;
        if(event.getRepeatCount() != 0) return true;
        int action = event.getAction();
        if(action == KeyEvent.ACTION_MULTIPLE) return true;
        // Ignore the cancelled up events. They occur when the user switches layouts.
        // In accordance with https://developer.android.com/reference/android/view/KeyEvent#FLAG_CANCELED
        if(action == KeyEvent.ACTION_UP &&
                (event.getFlags() & KeyEvent.FLAG_CANCELED) != 0) return true;

        //Sometimes, key events comes from SOME keys of the software keyboard
        //Even weirder, is is unknown why a key or another is selected to trigger a keyEvent
        if((event.getFlags() & KeyEvent.FLAG_SOFT_KEYBOARD) == KeyEvent.FLAG_SOFT_KEYBOARD){
            if(eventKeycode == KeyEvent.KEYCODE_ENTER) return true; //We already listen to it.
            touchCharInput.dispatchKeyEvent(event);
            return true;
        }

        //Sometimes, key events may come from the mouse
        if(event.getDevice() != null
                && ( (event.getSource() & InputDevice.SOURCE_MOUSE_RELATIVE) == InputDevice.SOURCE_MOUSE_RELATIVE
                ||   (event.getSource() & InputDevice.SOURCE_MOUSE) == InputDevice.SOURCE_MOUSE)  ){

            if(eventKeycode == KeyEvent.KEYCODE_BACK){
                sendMouseButton(MotionEvent.BUTTON_SECONDARY, event.getAction() == KeyEvent.ACTION_DOWN);
                return true;
            }
        }

        if(Gamepad.isGamepadEvent(event)){
            if(Platform.getPlatformGamepad() != null && Platform.getPlatformGamepad().shouldOverride()){
                Platform.getPlatformGamepad().sendKeyEvent(event);
                return true;
            }
            if(mGamepadHandler == null) createGamepad(event.getDevice());

            mInputManager.handleKeyEventInput(getContext(), event, mGamepadHandler);
            return true;
        }

        CallbackBridge.setModifiers(event);
        char codepoint = action == KeyEvent.ACTION_DOWN ? (char) event.getUnicodeChar(event.getMetaState()) : 0;
        PLATFORM.sendKeyEvent(eventKeycode, action == KeyEvent.ACTION_DOWN ? 1 : 0, CallbackBridge.getCurrentMods(), codepoint);

        // Some events will be generated an infinite number of times when no consumed
        return (event.getFlags() & KeyEvent.FLAG_FALLBACK) == KeyEvent.FLAG_FALLBACK;
    }

    /** Called when the size need to be set at any point during the surface lifecycle **/
    public void refreshSize(){
        refreshSize(false);
    }

    /** Same as refreshSize, but allows you to force an immediate size update **/
    public void refreshSize(boolean immediate) {
        if(isInLayout() && !immediate) {
            post(this::refreshSize);
            return;
        }
        // Use the width and height of the View instead of display dimensions to avoid
        // getting squiched/stretched due to inconsistencies between the layout and
        // screen dimensions.
        int newWidth = Tools.getDisplayFriendlyRes(getWidth(), LauncherPreferences.PREF_SCALE_FACTOR);
        int newHeight = Tools.getDisplayFriendlyRes(getHeight(), LauncherPreferences.PREF_SCALE_FACTOR);
        if (newHeight < 1 || newWidth < 1) {
            Log.e("MGLSurface", String.format("Impossible resolution : %dx%d", newWidth, newHeight));
            return;
        }
        windowWidth = newWidth;
        windowHeight = newHeight;
        windowRate = mSurface.getDisplay().getRefreshRate();
        if(mSurface == null){
            Log.w("MGLSurface", "Attempt to refresh size on null surface");
            return;
        }
        MojoExec.setDisplayParams(windowWidth, windowHeight, mSurface.getDisplay().getRefreshRate());
        mSurfaceProvider.updateSize();
    }

    private void realStart(){
        // Initial size set. Request immedate refresh, otherwise the initial width and height for the game
        // may be broken/unknown.
        refreshSize(true);

        //Load Minecraft options:
        MCOptionUtils.set("fullscreen", "off");
        MCOptionUtils.set("overrideWidth", String.valueOf(windowWidth));
        MCOptionUtils.set("overrideHeight", String.valueOf(windowHeight));
        MCOptionUtils.save();
        getMcScale();

        new Thread(() -> {
            try {
                // Wait until the listener is attached
                synchronized(mSurfaceReadyListenerLock) {
                    if(mSurfaceReadyListener == null) mSurfaceReadyListenerLock.wait();
                }

                mSurfaceReadyListener.isReady();
            } catch (Throwable e) {
                Tools.showError(getContext(), e, true);
            }
        }, "JVM Main thread").start();
    }

    @Override
    public void onGrabState(boolean isGrabbing) {
        if(mLastGrabState != isGrabbing) {
            Log.i("MGLSurface", "Grabbing state changed! " + mLastGrabState + " -> " + isGrabbing);
            mCurrentTouchProcessor.cancelPendingActions();
            mCurrentTouchProcessor = pickEventProcessor(isGrabbing);
            mLastGrabState = isGrabbing;
        }
    }

    private TouchEventProcessor pickEventProcessor(boolean isGrabbing) {
        return isGrabbing ? mIngameProcessor : mInGUIProcessor;
    }

    private void updateGrabState(boolean isGrabbing) {

    }

    @Override
    public void onSurfaceAvailable(Surface surface) {
        Platform.addGrabListener(this);
        Platform.setPendingSurface(surface);
        if(mRefreshOnly) return;
        realStart();
        mRefreshOnly = true;
    }

    @Override
    public void onSurfaceResized() {
        if(PLATFORM != null)
            PLATFORM.surfaceUpdated();
    }

    @Override
    public void onSurfaceDestroyed() {
        if(PLATFORM != null)
            PLATFORM.surfaceDestroyed();
    }

    @Override
    public void onEnableGamepad() {
        post(()->{
            if(mGamepadHandler != null && mGamepadHandler instanceof Gamepad) {
                ((Gamepad)mGamepadHandler).removeSelf();
            }
            // Force gamepad recreation on next event
            mGamepadHandler = null;
        });
    }

    /** A small interface called when the listener is ready for the first time */
    public interface SurfaceReadyListener {
        void isReady();
    }

    public void setSurfaceReadyListener(SurfaceReadyListener listener){
        synchronized (mSurfaceReadyListenerLock) {
            mSurfaceReadyListener = listener;
            mSurfaceReadyListenerLock.notifyAll();
        }
    }
    public static int getWindowWidth(){
        return windowWidth;
    }
    public static int getWindowHeight(){
        return windowHeight;
    }
    public static float getWindowRate(){
        return windowRate;
    }
}

File: app_pojavlauncher/src/main/java/net/kdt/pojavlaunch/LauncherGLSurface.kt
package net.kdt.pojavlaunch

import net.kdt.pojavlaunch.MainActivity.Companion.touchCharInput
import net.kdt.pojavlaunch.utils.MCOptionUtils.getMcScale
import net.kdt.pojavlaunch.CallbackBridge.sendMouseButton
import net.kdt.pojavlaunch.CallbackBridge.windowHeight
import net.kdt.pojavlaunch.CallbackBridge.windowWidth

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.util.Log
import android.view.InputDevice
import android.view.KeyEvent
import android.view.MotionEvent
import android.view.Surface
import android.view.View
import android.view.ViewGroup

import androidx.annotation.RequiresApi

import net.kdt.pojavlaunch.customcontrols.ControlLayout
import net.kdt.pojavlaunch.customcontrols.gamepad.DefaultDataProvider
import net.kdt.pojavlaunch.customcontrols.gamepad.Gamepad
import net.kdt.pojavlaunch.customcontrols.gamepad.DirectGamepad
import net.kdt.pojavlaunch.customcontrols.mouse.AndroidPointerCapture
import net.kdt.pojavlaunch.customcontrols.mouse.InGUIEventProcessor
import net.kdt.pojavlaunch.customcontrols.mouse.InGameEventProcessor
import net.kdt.pojavlaunch.customcontrols.mouse.TouchEventProcessor
import net.kdt.pojavlaunch.prefs.LauncherPreferences
import net.kdt.pojavlaunch.render.SurfaceProvider
import net.kdt.pojavlaunch.render.SurfaceViewSurfaceProvider
import net.kdt.pojavlaunch.render.TextureViewSurfaceProvider
import net.kdt.pojavlaunch.utils.JREUtils
import net.kdt.pojavlaunch.utils.MCOptionUtils

import fr.spse.gamepad_remapper.GamepadHandler
import fr.spse.gamepad_remapper.RemapperManager
import fr.spse.gamepad_remapper.RemapperView
import git.artdeell.dnbootstrap.glfw.GLFW
import git.artdeell.dnbootstrap.glfw.GamepadEnableHandler
import git.artdeell.dnbootstrap.glfw.GrabListener

class LauncherGLSurface : View, GrabListener, GamepadEnableHandler, SurfaceProvider.SurfaceCallback {
    private var mGamepadHandler: GamepadHandler? = null
    private val mInputManager = RemapperManager(context, RemapperView.Builder(null)
        .remapA(true)
        .remapB(true)
        .remapX(true)
        .remapY(true)
        .remapLeftJoystick(true)
        .remapRightJoystick(true)
        .remapStart(true)
        .remapSelect(true)
        .remapLeftShoulder(true)
        .remapRightShoulder(true)
        .remapLeftTrigger(true)
        .remapRightTrigger(true)
        .remapDpad(true))

    private val mSensitivityFactor = 1.4 * (1080f / Tools.getDisplayMetrics(context as Activity).heightPixels)
    private val mSurfaceProvider = if (LauncherPreferences.PREF_USE_ALTERNATE_SURFACE) SurfaceViewSurfaceProvider() else TextureViewSurfaceProvider()
    private var mRefreshOnly = true
    var mSurfaceReadyListener: SurfaceReadyListener? = null
    val mSurfaceReadyListenerLock = Any()
    var mSurface: View? = null

    private val mIngameProcessor = InGameEventProcessor(this, mSensitivityFactor)
    private val mInGUIProcessor = InGUIEventProcessor(this)
    private var mCurrentTouchProcessor: TouchEventProcessor = mInGUIProcessor
    private var mPointerCapture: AndroidPointerCapture? = null
    private var mTouchpad: View? = null
    private var mLastGrabState = false

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attributeSet: AttributeSet?) : super(context, attributeSet) {
        isFocusable = true
        GLFW.setGamepadEnableHandler(this)
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private fun setUpPointerCapture() {
        mPointerCapture?.detach()
        mPointerCapture = AndroidPointerCapture(mTouchpad!!, this)
    }

    fun start(isAlreadyRunning: Boolean, touchpad: View) {
        mTouchpad = touchpad
        if (Tools.isAndroid8OrHigher()) setUpPointerCapture()
        mInGUIProcessor.setAbstractTouchpad(touchpad)
        mRefreshOnly = isAlreadyRunning
        mSurface = mSurfaceProvider.create(context, this)
        (parent as ViewGroup).addView(mSurface)
    }

    @Suppress("accessibility")
    override fun onTouchEvent(e: MotionEvent): Boolean {
        if ((parent as ControlLayout).getModifiable()) return false

        for (i in 0 until e.pointerCount) {
            val toolType = e.getToolType(i)
            if (toolType == MotionEvent.TOOL_TYPE_MOUSE) {
                if (Tools.isAndroid8OrHigher() && mPointerCapture != null) {
                    mPointerCapture!!.handleAutomaticCapture()
                    return true
                }
            } else if (toolType != MotionEvent.TOOL_TYPE_STYLUS) continue

            if (GLFW.isGrabbing()) return false
            GLFW.cursorX = e.getX(i) / width.toDouble()
            GLFW.cursorY = e.getY(i) / height.toDouble()
            GLFW.sendMousePos()
            return true
        }
        if (mIngameProcessor == null || mInGUIProcessor == null) return true
        val ret = mCurrentTouchProcessor.processTouchEvent(e)
        if (LauncherPreferences.PREF_KEYBOARD_AUTOPANNING && MainActivity.mImeHeight > 0) {
            val translationY = Tools.getTranslationFromCursorY(
                (GLFW.cursorY * mSurface!!.height + 100).toInt(),
                mSurface!!.height,
                MainActivity.mImeHeight,
                0
            )
            if (MainActivity.mForceFullPanning) {
                mSurface!!.animate().setDuration(100).translationY(-translationY.toFloat()).start()
                mTouchpad!!.animate().setDuration(100).translationY(-translationY.toFloat()).start()
                MainActivity.mForceFullPanning = false
            } else {
                mSurface!!.translationY = -translationY.toFloat()
                mTouchpad!!.translationY = -translationY.toFloat()
            }
        }
        return ret
    }

    private fun createGamepad(inputDevice: InputDevice) {
        mGamepadHandler = if (GLFW.gamepadButtonBuffer != null) {
            GLFW.nativeNotifyGamepadConnected()
            DirectGamepad()
        } else {
            Gamepad(inputDevice, DefaultDataProvider.INSTANCE, mTouchpad)
        }
    }

    @SuppressLint("NewApi")
    override fun dispatchGenericMotionEvent(event: MotionEvent): Boolean {
        var mouseCursorIndex = -1

        if (Gamepad.isGamepadEvent(event)) {
            if (mGamepadHandler == null) createGamepad(event.device!!)
            mInputManager.handleMotionEventInput(context, event, mGamepadHandler)
            return true
        }

        for (i in 0 until event.pointerCount) {
            if (event.getToolType(i) != MotionEvent.TOOL_TYPE_MOUSE && event.getToolType(i) != MotionEvent.TOOL_TYPE_STYLUS) continue
            mouseCursorIndex = i
            break
        }
        if (mouseCursorIndex == -1) return false

        updateGrabState(GLFW.isGrabbing())

        when (event.actionMasked) {
            MotionEvent.ACTION_HOVER_MOVE -> {
                GLFW.cursorX = event.getX(mouseCursorIndex) / width.toDouble()
                GLFW.cursorY = event.getY(mouseCursorIndex) / height.toDouble()
                GLFW.sendMousePos()
                return true
            }
            MotionEvent.ACTION_SCROLL -> {
                CallbackBridge.sendScroll(event.getAxisValue(MotionEvent.AXIS_HSCROLL).toDouble(), event.getAxisValue(MotionEvent.AXIS_VSCROLL).toDouble())
                return true
            }
            MotionEvent.ACTION_BUTTON_PRESS -> return sendMouseButtonUnconverted(event.actionButton, true)
            MotionEvent.ACTION_BUTTON_RELEASE -> return sendMouseButtonUnconverted(event.actionButton, false)
            else -> return false
        }
    }

    fun processKeyEvent(event: KeyEvent): Boolean {
        val eventKeycode = event.keyCode
        if (eventKeycode == KeyEvent.KEYCODE_UNKNOWN) return true
        if (eventKeycode == KeyEvent.KEYCODE_VOLUME_DOWN) return false
        if (eventKeycode == KeyEvent.KEYCODE_VOLUME_UP) return false
        if (event.repeatCount != 0) return true
        val action = event.action
        if (action == KeyEvent.ACTION_MULTIPLE) return true
        if (action == KeyEvent.ACTION_UP &&
            (event.flags and KeyEvent.FLAG_CANCELED) != 0) return true

        if ((event.flags and KeyEvent.FLAG_SOFT_KEYBOARD) == KeyEvent.FLAG_SOFT_KEYBOARD) {
            if (eventKeycode == KeyEvent.KEYCODE_ENTER) return true
            touchCharInput!!.dispatchKeyEvent(event)
            return true
        }

        if (event.device != null
            && (((event.source and InputDevice.SOURCE_MOUSE_RELATIVE) == InputDevice.SOURCE_MOUSE_RELATIVE
                    || (event.source and InputDevice.SOURCE_MOUSE) == InputDevice.SOURCE_MOUSE))) {
            if (eventKeycode == KeyEvent.KEYCODE_BACK) {
                sendMouseButton(LwjglGlfwKeycode.GLFW_MOUSE_BUTTON_RIGHT, event.action == KeyEvent.ACTION_DOWN)
                return true
            }
        }

        if (Gamepad.isGamepadEvent(event)) {
            if (mGamepadHandler == null) createGamepad(event.device!!)
            mInputManager.handleKeyEventInput(context, event, mGamepadHandler)
            return true
        }

        CallbackBridge.setModifiers(event)
        val codepoint = if (action == KeyEvent.ACTION_DOWN) event.getUnicodeChar(event.metaState).toChar() else 0.toChar()
        GLFW.sendRawKeyEvent(eventKeycode, if (action == KeyEvent.ACTION_DOWN) 1 else 0, CallbackBridge.currentMods, codepoint)

        return (event.flags and KeyEvent.FLAG_FALLBACK) == KeyEvent.FLAG_FALLBACK
    }

    companion object {
        fun sendMouseButtonUnconverted(button: Int, status: Boolean): Boolean {
            val glfwButton = when (button) {
                MotionEvent.BUTTON_PRIMARY -> LwjglGlfwKeycode.GLFW_MOUSE_BUTTON_LEFT
                MotionEvent.BUTTON_TERTIARY -> LwjglGlfwKeycode.GLFW_MOUSE_BUTTON_MIDDLE
                MotionEvent.BUTTON_SECONDARY -> LwjglGlfwKeycode.GLFW_MOUSE_BUTTON_RIGHT
                else -> -256
            }
            if (glfwButton == -256) return false
            sendMouseButton(glfwButton, status)
            return true
        }
    }

    fun refreshSize() {
        refreshSize(false)
    }

    fun refreshSize(immediate: Boolean) {
        if (isInLayout && !immediate) {
            post { refreshSize() }
            return
        }
        val newWidth = Tools.getDisplayFriendlyRes(width, LauncherPreferences.PREF_SCALE_FACTOR)
        val newHeight = Tools.getDisplayFriendlyRes(height, LauncherPreferences.PREF_SCALE_FACTOR)
        if (newHeight < 1 || newWidth < 1) {
            Log.e("MGLSurface", String.format("Impossible resolution : %dx%d", newWidth, newHeight))
            return
        }
        windowWidth = newWidth
        windowHeight = newHeight
        if (mSurface == null) {
            Log.w("MGLSurface", "Attempt to refresh size on null surface")
            return
        }
        JREUtils.configureRenderspecDisplay(windowWidth, windowHeight, mSurface!!.display.refreshRate.toInt())
        mSurfaceProvider.updateSize()
    }

    private fun realStart() {
        refreshSize(true)
        MCOptionUtils.set("fullscreen", "off")
        MCOptionUtils.set("overrideWidth", windowWidth.toString())
        MCOptionUtils.set("overrideHeight", windowHeight.toString())
        MCOptionUtils.save()
        getMcScale()

        Thread {
            try {
                synchronized(mSurfaceReadyListenerLock) {
                    if (mSurfaceReadyListener == null) mSurfaceReadyListenerLock.wait()
                }
                mSurfaceReadyListener!!.isReady()
            } catch (e: Throwable) {
                Tools.showError(context, e, true)
            }
        }.start()
    }

    override fun onGrabState(isGrabbing: Boolean) {
        post { updateGrabState(isGrabbing) }
    }

    private fun pickEventProcessor(isGrabbing: Boolean): TouchEventProcessor {
        return if (isGrabbing) mIngameProcessor else mInGUIProcessor
    }

    private fun updateGrabState(isGrabbing: Boolean) {
        if (mLastGrabState != isGrabbing) {
            mCurrentTouchProcessor.cancelPendingActions()
            mCurrentTouchProcessor = pickEventProcessor(isGrabbing)
            mLastGrabState = isGrabbing
        }
    }

    override fun onSurfaceAvailable(surface: Surface) {
        GLFW.nativeSurfaceCreated(surface)
        if (mRefreshOnly) return
        realStart()
        mRefreshOnly = true
    }

    override fun onSurfaceResized() {
        GLFW.nativeSurfaceUpdated()
    }

    override fun onSurfaceDestroyed() {
        GLFW.nativeSurfaceDestroyed()
    }

    override fun onEnableGamepad() {
        post {
            if (mGamepadHandler != null && mGamepadHandler is Gamepad) {
                (mGamepadHandler as Gamepad).removeSelf()
            }
            mGamepadHandler = null
        }
    }

    interface SurfaceReadyListener {
        fun isReady()
    }

    fun setSurfaceReadyListener(listener: SurfaceReadyListener) {
        synchronized(mSurfaceReadyListenerLock) {
            mSurfaceReadyListener = listener
            mSurfaceReadyListenerLock.notifyAll()
        }
    }
}
